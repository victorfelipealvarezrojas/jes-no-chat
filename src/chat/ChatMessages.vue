<template>
  <div ref="chatRef" class="flex-1 overflow-y-auto p-4">
    <div class="flex flex-col space-y-2">
      <ChatBubble
        v-for="message in messages"
        :key="message.id"
        :its-mine="message.itsMine"
        :message="message.message"
        :image="message.image"
      />
    </div>
  </div>
</template>

<script setup lang="ts">
import type { ChatMessage } from '@/interfaces/chat-messages.interfaces'
import ChatBubble from './ChatBubble.vue'
import { ref, watch } from 'vue'

interface Props {
  messages: ChatMessage[]
}

const props = defineProps<Props>()

const chatRef = ref<HTMLDivElement | null>(null)

watch(
  () => props.messages.values(),
  () => {
    setTimeout(() => {
      chatRef.value?.scrollTo({
        top: chatRef.value.scrollHeight,
        behavior: 'smooth',
      })
    }, 100)
  },
)
</script>
