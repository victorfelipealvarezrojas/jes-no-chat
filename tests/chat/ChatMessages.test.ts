import { mount } from '@vue/test-utils'
import ChatMessages from '@/chat/ChatMessages.vue'
import type { ChatMessage } from '@/interfaces/chat-messages.interfaces'

const messages: ChatMessage[] = [
  { id: 1, message: 'Hello shiao', itsMine: true },
  { id: 2, message: 'World plop', itsMine: false },
  { id: 3, message: 'Hello World', itsMine: true },
]

describe('ChatMessages', () => {
  test('snapshot correctly ', () => {
    const wrapper = mount(ChatMessages, {
      props: { messages },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('renders messages correctly', () => {
    const wrapper = mount(ChatMessages, {
      props: { messages },
    })

    const chatBubble = wrapper.findAllComponents({ name: 'ChatBubble' })
    expect(chatBubble).toHaveLength(3)
  })

  test('scrolls to bottom when messages are updated', async () => {
    const wrapper = mount(ChatMessages, {
      props: { messages },
    })

    const scrollmock = vi.fn()
    const refScroll = wrapper.vm.$refs.chatRef as any
    refScroll.scrollTo = scrollmock

    await wrapper.setProps({
      messages: [
        ...messages,
        { id: 4, message: 'Hello shiao 4', itsMine: true },
        { id: 5, message: 'World plop 5', itsMine: false },
        { id: 6, message: 'Hello World 6', itsMine: true },
        { id: 7, message: 'Hello shiao 7', itsMine: true },
      ],
    })

    await new Promise((resolve) => setTimeout(resolve, 150))

    expect(scrollmock).toHaveBeenCalled()
  })
})
