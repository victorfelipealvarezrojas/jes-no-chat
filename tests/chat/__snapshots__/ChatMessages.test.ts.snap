// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChatMessages > snapshot correctly  1`] = `
"<div class="flex-1 overflow-y-auto p-4">
  <div class="flex flex-col space-y-2">
    <!-- Messages go here -->
    <!-- Example Message -->
    <div class="flex justify-end">
      <div class="bg-blue-200 text-black p-2 rounded-lg max-w-xs">Hello shiao</div>
    </div><!-- Messages go here -->
    <!-- Example Message -->
    <!-- Example Received Message -->
    <div class="flex">
      <div class="bg-gray-300 text-black p-2 rounded-lg max-w-xs"><span class="capitalize">World plop</span>
        <!--v-if-->
      </div>
    </div><!-- Messages go here -->
    <!-- Example Message -->
    <div class="flex justify-end">
      <div class="bg-blue-200 text-black p-2 rounded-lg max-w-xs">Hello World</div>
    </div>
  </div>
</div>"
`;
